src/lib/operators/join_sort_merge.cpp

```diff
@@ -247,21 +257,62 @@ class JoinSortMerge::JoinSortMergeImpl : public AbstractJoinOperatorImpl {
   /**
   * Emits a combination of a left row id and a right row id to the join output.
   **/
-  void _emit_combination(size_t output_cluster, RowID left, RowID right) {
-    _output_pos_lists_left[output_cluster]->push_back(left);
-    _output_pos_lists_right[output_cluster]->push_back(right);
+  void _emit_combination(size_t output_cluster, RowID left_row_id, RowID right_row_id) {
+    _output_pos_lists_left[output_cluster]->push_back(left_row_id);
+    _output_pos_lists_right[output_cluster]->push_back(right_row_id);
   }
 
   /**
-  * Emits all the combinations of row ids from the left table range and the right table range to the join output.
-  * I.e. the cross product of the ranges is emitted.
-  **/
-  void _emit_all_combinations(size_t output_cluster, TableRange left_range, TableRange right_range) {
-    left_range.for_every_row_id(_sorted_left_table, [&](RowID left_row_id) {
-      right_range.for_every_row_id(_sorted_right_table, [&](RowID right_row_id) {
-        _emit_combination(output_cluster, left_row_id, right_row_id);
+    * Emits all the combinations of row ids from the left table range and the right table range to the join output
+    * where also the secondary predicates are satisfied.
+    **/
+  void _emit_qualified_combinations(size_t output_cluster, TableRange left_range, TableRange right_range) {
```

Comment:
WIP!

Reply:
you are welcome to be more specific

Reply:
we will distinguish the different cases so we can avoid using the map for left and right join. Therefore just the full outer join is the expensive one where we use the map.